brighthive/data-resource-api
# Feat/many to many

## Overview

Adds the ability to create resources with many to many (M:N) relationships and to view the data. This allows our data trusts to be created that have M:N relationships. For instance many providers can have many credentials.

1. Creates association table from descriptor file
[ORM Factory](https://github.com/brighthive/data-resource-api/blob/83c1f204cf41c530c193af6b0d2720eb88f4c806/data_resource_api/factories/orm_factory.py#L159-L174) now creates the association table and stores it to a static class called the (junction table holder)[https://github.com/brighthive/data-resource-api/blob/feat/many-to-many/data_resource_api/app/junc_holder.py].

2. Creates m:n route
Extended the [data resource factory](https://github.com/brighthive/data-resource-api/blob/83c1f204cf41c530c193af6b0d2720eb88f4c806/data_resource_api/factories/data_resource_factory.py#L56-L81) to create the M:N routes.

3. Versioned Resource refactored
[Versioned Resource](https://github.com/brighthive/data-resource-api/blob/feat/many-to-many/data_resource_api/api/core/versioned_resource.py) has been split into three classes. A parent class and two children classes; one for singular resources and one for M:N resources.

4. Create a resource with a many to many relationship
The [insert one method](https://github.com/brighthive/data-resource-api/blob/83c1f204cf41c530c193af6b0d2720eb88f4c806/data_resource_api/api/v1_0_0/resource_handler.py#L241-L313) has been extended to support M:N relationships on object creation. When a resource is created (via POST) with a child relationship the code will add the relation to the junction table.

5. View the many to many data by doing `GET /parent/id/child`
By performing a GET to the appropriate route you will hit the [get_many_one method](https://github.com/brighthive/data-resource-api/blob/83c1f204cf41c530c193af6b0d2720eb88f4c806/data_resource_api/api/v1_0_0/resource_handler.py#L380-L405) via the [VersionedResourceMany](https://github.com/brighthive/data-resource-api/blob/83c1f204cf41c530c193af6b0d2720eb88f4c806/data_resource_api/api/core/versioned_resource.py#L38). This will query the junction table and return the requested relationship data found in the relationship.

### Demo

![manytomany](https://user-images.githubusercontent.com/31484824/67796443-49d29f80-fa3d-11e9-8198-e3f8de83e254.gif)

### Notes

Originally I attempted to perform M:N operations using SQLAlchemy's many-to-many ORM relationship methods but that proved to be very complicated so I chose to instead perform the operations on the association table directly.

## Testing Instructions

 * Start after checking out this branch.
 * To be safe ensure you have killed any running docker images related to Data Resource API.
 * Create the data resource and run it with;
```bash
docker build -t brighthive/data-resource-api:1.0.0-alpha .
docker-compose up
```
 * Allow the data resource to initialize for a minute or two. You may see errors occur at first.
 * Open POSTMAN.
 * Create two credentials by performing a POST to `http://localhost:8000/credentials` with the following JSON body:
```json
{
	"credential_name": "credential test"
}
```

 * Create a program with two credentials by performing a POST to `http://localhost:8000/programs` with the following JSON body:
```json
{
	"program_name": "aprogram 1",
	"program_code": 1,
	"program_description": ".",
	"program_status": ".",
	"program_fees": 1,
	"eligibility_criteria": ".",
	"program_url": ".",
	"credentials": [1,2]
}
```

 * Get the data by performing `GET http://localhost:8000/programs/1/credentials`

 * You should see
```
{
    "credentials": [
        1,
        2
    ]
}
```

Handles https://trello.com/c/MzCjBjvc/147-2-dr-mn-get and https://trello.com/c/95mQmdAi/143-3-dr-resource-post-with-many-to-many
